File: scripts/ml-utils/ml-client.js
/**
 * ML Tools JavaScript Client
 * A JavaScript/Node.js client library for interacting with the ML Tools API
 */

class MLToolsClient {
  /**
   * Initialize the ML Tools client
   * @param {string} baseUrl - Base URL of the ML Tools API
   * @param {string} apiKey - Optional API key for authentication
   */
  constructor(baseUrl, apiKey = null) {
    this.baseUrl = baseUrl.replace(/\/$/, '');
    this.apiKey = apiKey;
  }

  /**
   * Make an HTTP request
   * @private
   */
  async _request(method, path, data = null) {
    const url = `${this.baseUrl}${path}`;
    const options = {
      method,
      headers: {
        'Content-Type': 'application/json'
      }
    };

    if (this.apiKey) {
      options.headers['Authorization'] = `Bearer ${this.apiKey}`;
    }

    if (data) {
      options.body = JSON.stringify(data);
    }

    const response = await fetch(url, options);
    
    if (!response.ok) {
      throw new Error(`HTTP ${response.status}: ${response.statusText}`);
    }

    return await response.json();
  }

  /**
   * Run a single prediction
   * @param {string} modelId - ID of the model to use
   * @param {Object} inputData - Input data for prediction
   * @returns {Promise<Object>} Prediction result
   */
  async predict(modelId, inputData) {
    return await this._request('POST', '/api/v1/ml/predict', {
      modelId,
      input: inputData
    });
  }

  /**
   * Run batch predictions
   * @param {string} modelId - ID of the model to use
   * @param {Array<Object>} inputs - List of input data for predictions
   * @returns {Promise<Object>} Batch prediction results
   */
  async predictBatch(modelId, inputs) {
    return await this._request('POST', '/api/v1/ml/predict-batch', {
      modelId,
      inputs
    });
  }

  /**
   * Get list of all active models
   * @returns {Promise<Array<Object>>} List of active models
   */
  async getActiveModels() {
    const result = await this._request('GET', '/api/v1/ml/models/active');
    return result.models;
  }

  /**
   * Get details about a specific model
   * @param {string} modelId - ID of the model
   * @returns {Promise<Object>} Model details
   */
  async getModel(modelId) {
    return await this._request('GET', `/api/v1/ml/models/${modelId}`);
  }

  /**
   * Check API health
   * @returns {Promise<Object>} Health status
   */
  async healthCheck() {
    return await this._request('GET', '/api/v1/ml/health');
  }
}

class ModelManager {
  /**
   * Helper class for managing ML models
   * @param {MLToolsClient} client - ML Tools client instance
   */
  constructor(client) {
    this.client = client;
  }

  /**
   * Print list of all active models
   */
  async listModels() {
    const models = await this.client.getActiveModels();

    console.log('\n' + '='.repeat(80));
    console.log(`Active Models (${models.length})`);
    console.log('='.repeat(80) + '\n');

    models.forEach((model, i) => {
      console.log(`${i + 1}. ${model.name} (v${model.version})`);
      console.log(`   ID: ${model.id}`);
      console.log(`   Type: ${model.type}`);
      console.log(`   Input Shape: ${model.inputShape || 'N/A'}`);
      console.log(`   Output Shape: ${model.outputShape || 'N/A'}`);
      console.log();
    });
  }

  /**
   * Print detailed information about a model
   * @param {string} modelId - ID of the model
   */
  async showModelDetails(modelId) {
    const model = await this.client.getModel(modelId);

    console.log('\n' + '='.repeat(80));
    console.log(`Model Details: ${model.name}`);
    console.log('='.repeat(80) + '\n');

    console.log(`ID: ${model.id}`);
    console.log(`Version: ${model.version}`);
    console.log(`Type: ${model.type}`);
    console.log(`Framework: ${model.framework || 'N/A'}`);
    console.log(`Description: ${model.description || 'N/A'}`);
    console.log(`Input Shape: ${model.inputShape || 'N/A'}`);
    console.log(`Output Shape: ${model.outputShape || 'N/A'}`);
    console.log(`Active: ${model.active}`);

    if (model.metrics) {
      console.log('\nMetrics:');
      Object.entries(model.metrics).forEach(([key, value]) => {
        console.log(`  ${key}: ${value}`);
      });
    }

    console.log(`\nCreated: ${model.createdAt || 'N/A'}`);
    console.log(`Updated: ${model.updatedAt || 'N/A'}`);
    console.log();
  }
}

class PredictionRunner {
  /**
   * Helper class for running predictions
   * @param {MLToolsClient} client - ML Tools client instance
   */
  constructor(client) {
    this.client = client;
  }

  /**
   * Run a prediction and optionally print results
   * @param {string} modelId - ID of the model to use
   * @param {Object} inputData - Input data for prediction
   * @param {boolean} verbose - Whether to print results
   * @returns {Promise<Object>} Prediction result
   */
  async runPrediction(modelId, inputData, verbose = true) {
    const result = await this.client.predict(modelId, inputData);

    if (verbose) {
      console.log('\n' + '='.repeat(80));
      console.log('Prediction Result');
      console.log('='.repeat(80) + '\n');

      console.log(`Model: ${result.modelName} (v${result.modelVersion})`);
      console.log(`Timestamp: ${result.timestamp}`);
      console.log('\nPrediction:');
      console.log(`  Result: ${result.prediction.result}`);
      console.log(`  Confidence: ${(result.prediction.confidence * 100).toFixed(2)}%`);
      console.log(`  Processing Time: ${result.prediction.processingTime.toFixed(3)}s`);
      console.log();
    }

    return result;
  }

  /**
   * Run batch predictions and optionally print results
   * @param {string} modelId - ID of the model to use
   * @param {Array<Object>} inputs - List of input data for predictions
   * @param {boolean} verbose - Whether to print results
   * @returns {Promise<Object>} Batch prediction results
   */
  async runBatchPrediction(modelId, inputs, verbose = true) {
    const result = await this.client.predictBatch(modelId, inputs);

    if (verbose) {
      console.log('\n' + '='.repeat(80));
      console.log('Batch Prediction Results');
      console.log('='.repeat(80) + '\n');

      console.log(`Model: ${result.modelName} (v${result.modelVersion})`);
      console.log(`Batch Size: ${result.batchSize}`);
      console.log(`Timestamp: ${result.timestamp}`);
      console.log('\nPredictions:');

      result.predictions.forEach(pred => {
        console.log(`  [${pred.index}] Result: ${pred.result} | ` +
                   `Confidence: ${(pred.confidence * 100).toFixed(2)}%`);
      });
      console.log();
    }

    return result;
  }
}

// Example usage
async function main() {
  // Initialize client
  const client = new MLToolsClient('http://localhost:3000');

  // Check API health
  try {
    const health = await client.healthCheck();
    console.log(`API Status: ${health.status}`);
    console.log(`Version: ${health.version}`);
  } catch (error) {
    console.error(`Error connecting to API: ${error.message}`);
    process.exit(1);
  }

  // List active models
  const manager = new ModelManager(client);
  await manager.listModels();

  // Example: Run a prediction (replace with actual model ID)
  // const runner = new PredictionRunner(client);
  // await runner.runPrediction(
  //   'your-model-id',
  //   { data: [1, 2, 3, 4, 5] }
  // );

  // Example: Run batch prediction
  // await runner.runBatchPrediction(
  //   'your-model-id',
  //   [
  //     { data: [1, 2, 3] },
  //     { data: [4, 5, 6] }
  //   ]
  // );
}

// Export for use as a module
if (typeof module !== 'undefined' && module.exports) {
  module.exports = {
    MLToolsClient,
    ModelManager,
    PredictionRunner
  };
}

// Run if executed directly
if (typeof require !== 'undefined' && require.main === module) {
  main().catch(console.error);
}

File: scripts/ml-utils/ml_client.py
#!/usr/bin/env python3
"""
ML Tools Python Client
A Python client library for interacting with the ML Tools API
"""

import requests
import json
from typing import Dict, List, Any, Optional
from datetime import datetime


class MLToolsClient:
    """Client for interacting with ML Tools API"""
    
    def __init__(self, base_url: str, api_key: Optional[str] = None):
        """
        Initialize the ML Tools client
        
        Args:
            base_url: Base URL of the ML Tools API (e.g., 'http://localhost:3000')
            api_key: Optional API key for authentication
        """
        self.base_url = base_url.rstrip('/')
        self.api_key = api_key
        self.session = requests.Session()
        
        if api_key:
            self.session.headers.update({'Authorization': f'Bearer {api_key}'})
    
    def predict(self, model_id: str, input_data: Dict[str, Any]) -> Dict[str, Any]:
        """
        Run a single prediction
        
        Args:
            model_id: ID of the model to use
            input_data: Input data for prediction
            
        Returns:
            Prediction result
        """
        url = f"{self.base_url}/api/v1/ml/predict"
        payload = {
            'modelId': model_id,
            'input': input_data
        }
        
        response = self.session.post(url, json=payload)
        response.raise_for_status()
        return response.json()
    
    def predict_batch(self, model_id: str, inputs: List[Dict[str, Any]]) -> Dict[str, Any]:
        """
        Run batch predictions
        
        Args:
            model_id: ID of the model to use
            inputs: List of input data for predictions
            
        Returns:
            Batch prediction results
        """
        url = f"{self.base_url}/api/v1/ml/predict-batch"
        payload = {
            'modelId': model_id,
            'inputs': inputs
        }
        
        response = self.session.post(url, json=payload)
        response.raise_for_status()
        return response.json()
    
    def get_active_models(self) -> List[Dict[str, Any]]:
        """
        Get list of all active models
        
        Returns:
            List of active models
        """
        url = f"{self.base_url}/api/v1/ml/models/active"
        response = self.session.get(url)
        response.raise_for_status()
        return response.json()['models']
    
    def get_model(self, model_id: str) -> Dict[str, Any]:
        """
        Get details about a specific model
        
        Args:
            model_id: ID of the model
            
        Returns:
            Model details
        """
        url = f"{self.base_url}/api/v1/ml/models/{model_id}"
        response = self.session.get(url)
        response.raise_for_status()
        return response.json()
    
    def health_check(self) -> Dict[str, Any]:
        """
        Check API health
        
        Returns:
            Health status
        """
        url = f"{self.base_url}/api/v1/ml/health"
        response = self.session.get(url)
        response.raise_for_status()
        return response.json()


class ModelManager:
    """Helper class for managing ML models"""
    
    def __init__(self, client: MLToolsClient):
        self.client = client
    
    def list_models(self) -> None:
        """Print list of all active models"""
        models = self.client.get_active_models()
        
        print(f"\n{'='*80}")
        print(f"Active Models ({len(models)})")
        print(f"{'='*80}\n")
        
        for i, model in enumerate(models, 1):
            print(f"{i}. {model['name']} (v{model['version']})")
            print(f"   ID: {model['id']}")
            print(f"   Type: {model['type']}")
            print(f"   Input Shape: {model.get('inputShape', 'N/A')}")
            print(f"   Output Shape: {model.get('outputShape', 'N/A')}")
            print()
    
    def show_model_details(self, model_id: str) -> None:
        """Print detailed information about a model"""
        model = self.client.get_model(model_id)
        
        print(f"\n{'='*80}")
        print(f"Model Details: {model['name']}")
        print(f"{'='*80}\n")
        
        print(f"ID: {model['id']}")
        print(f"Version: {model['version']}")
        print(f"Type: {model['type']}")
        print(f"Framework: {model.get('framework', 'N/A')}")
        print(f"Description: {model.get('description', 'N/A')}")
        print(f"Input Shape: {model.get('inputShape', 'N/A')}")
        print(f"Output Shape: {model.get('outputShape', 'N/A')}")
        print(f"Active: {model.get('active', False)}")
        
        if model.get('metrics'):
            print(f"\nMetrics:")
            for key, value in model['metrics'].items():
                print(f"  {key}: {value}")
        
        print(f"\nCreated: {model.get('createdAt', 'N/A')}")
        print(f"Updated: {model.get('updatedAt', 'N/A')}")
        print()


class PredictionRunner:
    """Helper class for running predictions"""
    
    def __init__(self, client: MLToolsClient):
        self.client = client
    
    def run_prediction(self, model_id: str, input_data: Dict[str, Any], 
                      verbose: bool = True) -> Dict[str, Any]:
        """
        Run a prediction and optionally print results
        
        Args:
            model_id: ID of the model to use
            input_data: Input data for prediction
            verbose: Whether to print results
            
        Returns:
            Prediction result
        """
        result = self.client.predict(model_id, input_data)
        
        if verbose:
            print(f"\n{'='*80}")
            print(f"Prediction Result")
            print(f"{'='*80}\n")
            
            print(f"Model: {result['modelName']} (v{result['modelVersion']})")
            print(f"Timestamp: {result['timestamp']}")
            print(f"\nPrediction:")
            print(f"  Result: {result['prediction']['result']}")
            print(f"  Confidence: {result['prediction']['confidence']:.2%}")
            print(f"  Processing Time: {result['prediction']['processingTime']:.3f}s")
            print()
        
        return result
    
    def run_batch_prediction(self, model_id: str, inputs: List[Dict[str, Any]], 
                            verbose: bool = True) -> Dict[str, Any]:
        """
        Run batch predictions and optionally print results
        
        Args:
            model_id: ID of the model to use
            inputs: List of input data for predictions
            verbose: Whether to print results
            
        Returns:
            Batch prediction results
        """
        result = self.client.predict_batch(model_id, inputs)
        
        if verbose:
            print(f"\n{'='*80}")
            print(f"Batch Prediction Results")
            print(f"{'='*80}\n")
            
            print(f"Model: {result['modelName']} (v{result['modelVersion']})")
            print(f"Batch Size: {result['batchSize']}")
            print(f"Timestamp: {result['timestamp']}")
            print(f"\nPredictions:")
            
            for pred in result['predictions']:
                print(f"  [{pred['index']}] Result: {pred['result']} | "
                      f"Confidence: {pred['confidence']:.2%}")
            print()
        
        return result


# Example usage
if __name__ == '__main__':
    # Initialize client
    client = MLToolsClient('http://localhost:3000')
    
    # Check API health
    try:
        health = client.health_check()
        print(f"API Status: {health['status']}")
        print(f"Version: {health['version']}")
    except Exception as e:
        print(f"Error connecting to API: {e}")
        exit(1)
    
    # List active models
    manager = ModelManager(client)
    manager.list_models()
    
    # Example: Run a prediction (replace with actual model ID)
    # runner = PredictionRunner(client)
    # result = runner.run_prediction(
    #     model_id='your-model-id',
    #     input_data={'data': [1, 2, 3, 4, 5]}
    # )
    
    # Example: Run batch prediction
    # result = runner.run_batch_prediction(
    #     model_id='your-model-id',
    #     inputs=[
    #         {'data': [1, 2, 3]},
    #         {'data': [4, 5, 6]}
    #     ]
    # )
